### DISCUSSION TITLE
Version.h isn't locally updated : how to improve ?

### DISCUSSION THREAD

**Discussion by 0penBrain** [FreeCAD/FreeCAD#9606] [Code contribution discussions]
Hi,

I have an issue that `src/Build/Version.h` isn't updated on my local git copy when I pull or switch branches.
And so version displayed in FC info is never correct. I have to manually delete the former file so it updated.

With limited skills in CMake, I tried to improve situation:
* One try using only targets there: https://github.com/0penBrain/FreeCAD/tree/ver_targets
* Another try using interface libraries there: https://github.com/0penBrain/FreeCAD/tree/ver_interface_lib

I would say it's better but still not perfect as I need to run 2 consecutive builds so everything is correctly updated.

I have the feeling that on 1st run, `Version.h` is updated but after CMake analyses dependencies so it decided to not rebuild `FreeCADApp` (as at the time of analyze, `Version.h` wasn't changed).
On 2nd run, as `Version.h` has been changed by previous run, dependency analysis works as expected and trigger `FreeCADApp` rebuild.

May first one confirm that current behavior is "faulty" ?
Could be only due to my environment.

If confirmed, may someone skilled at CMake give a hand ? @chennes @wwmayer 

To reproduce/test, the simplest is to have 2 identical branches and switch between them.

Thx

**Comment by luzpaz** [comment]
Related to https://github.com/FreeCAD/FreeCAD/issues/5945

**Comment by wwmayer** [comment]
> I have an issue that src/Build/Version.h isn't updated on my local git copy

Version.h is not part of the git copy but it's created in the build directory when running CMake. Whenever I pull the latest changes or switch to a branch and run the build process then Version.h always gets correctly updated.

**Reply by 0penBrain** [reply]
> Whenever I pull the latest changes or switch to a branch and run the build process then Version.h always gets correctly updated.

Not me. 😄 Lucky you. I use CMake/Ninja/CCache/Clang. Maybe one of these makes a difference. 😕

**Comment by berniev** [comment]
`CMake/Ninja/CCache/Clang`. Same here. version.h is set by build. Certainly the date and time are right. The version number I don't know if right or not.

**Comment by wwmayer** [comment]
> I use CMake/Ninja/CCache/Clang. Maybe one of these makes a difference.

I have tried ninja and this seems to make the difference. For the initial build Version.h is created but for builds after a commit or branch switch it only updates the file Version.h.out and leaves Version.h untouched. It looks like the custom command
```
    add_custom_command(
        OUTPUT  ${CMAKE_CURRENT_BINARY_DIR}/Version.h
        COMMAND ${CMAKE_COMMAND} -E copy_if_different ${CMAKE_CURRENT_BINARY_DIR}/Version.h.out
            ${CMAKE_CURRENT_BINARY_DIR}/Version.h
        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
        DEPENDS version_check)
```
fails to work for ninja builds.

**Comment by wwmayer** [comment]
You will find example code here: https://cmake.org/cmake/help/latest/command/add_custom_command.html#example-generating-files-for-multiple-targets that is quite similar to our use case. However, when doing it this way then Version.h won't be updated any more at all.

Doing some further experiments I figured out that for ninja this makes it partially work -- the Version.h is updated for every second call of `ninja fc_version`. However, the classical make now prints a warning about a circular dependency.
```
    add_custom_command(
        OUTPUT  ${CMAKE_CURRENT_BINARY_DIR}/Version.h
        COMMAND ${CMAKE_COMMAND} -E copy_if_different ${CMAKE_CURRENT_BINARY_DIR}/Version.h.out
            ${CMAKE_CURRENT_BINARY_DIR}/Version.h
        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
        DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/Version.h.out version_check)
```

**Reply by 0penBrain** [reply]
> Doing some further experiments I figured out that for ninja this makes it partially work -- the Version.h is updated for every second call of `ninja fc_version`

OK, so same as my 2 different tries in the OP. I'll try passing them to Make in the coming days to check if they also return cyclic dependency. 😉 Thx for the insight.

**Reply by luzpaz** [reply]
Any success ?

**Reply by 0penBrain** [reply]
Not get the time yet to test. Still on my todo list. ;)
